### DISCUSSION TITLE
Basic Hyperparameter Tuning for Time Series Models in PyCaret

### DISCUSSION THREAD

**Discussion by ngupta23** [pycaret/pycaret#1791] [FAQs]
# Step 1: Create a simple baseline model

Before we start to tune a model, we need to create a simple baseline model first. Please refer to **[this post](https://github.com/pycaret/pycaret/discussions/1757#discussion-3645992)** on how you can do that. We show a simple reduced regression model below which uses linear regression to model the monthly airline dataset. A complete notebook example for tuning can be **[found here](https://gist.github.com/ngupta23/5ea07c429625ed536aa7922e19e94ec4)**.

```python
model = exp.create_model("lr_cds_dt")
exp.plot_model(model)
```

![image](https://user-images.githubusercontent.com/33585645/139674457-b010410a-7e7e-4795-8518-0d4cc32d5ab8.png)

![ts_tune_model_1p2](https://user-images.githubusercontent.com/33585645/139825950-a607b3fb-a72d-4466-b200-f0d3d6450af5.png)

```python
#### What are the model hyper-parameters? ----
model
```

![image](https://user-images.githubusercontent.com/33585645/139673376-4db552ab-4e8a-4caf-aa81-b443f72c5a36.png)

The model uses a degree of 1 to detrend the data set before fitting. It also uses a seasonal_period of 1 by default with a window length of 10 (used to extract lagged features. i.e. uses up to 10 lags from the past).

# Step 2: Tune the model

## Random Grid Search

By default, `pycaret` performs a random grid search using a predefined internal grid. This default internal grid uses sensible defaults (such as `seasonal_period`) obtained during the `setup`.

```python
#### Tune the model (Default = Random Grid Search)
tuned_model_random = exp.tune_model(model)
exp.plot_model(tuned_model_random)
```
![image](https://user-images.githubusercontent.com/33585645/139674700-5ad268cb-56b0-4cd8-82fe-9d33157b2a97.png)

![ts_tune_model_1p3](https://user-images.githubusercontent.com/33585645/139826069-ba667197-e733-4b09-ab0f-532afd7cce85.png)

```python
#### What are the tuned model hyper-parameters? ----
tuned_model_random
```

![image](https://user-images.githubusercontent.com/33585645/139673809-b66c7185-3e9f-4ab3-b130-a3defa07b7eb.png)

As we can see, the tuned model has a much better performance than the original model (e.g. MAE reduced substantially). The tuned model uses a degree of 2 to detrend the data, seasonal period of 12 which seems reasonable for a monthly airline data and a window length (number of lags) of 23. 

One thing that you might be curious about is what was the search space used? This can be obtained easily using the `models()` call with `internal=True`. The column `Tune Distributions` provides the default random grid used for tuning. In this case, six (6) hyper parameters were tuned in a random manner. Note that

1. The limits of the search space are chosen carefully based on the time series characteristics. For example,
    - The seasonal period `sp` is chosen to be a choice between 12 and 24. This is based on the seasonality of 12 that is detected during the `setup` process. You may wonder why we also add a seasonality of 24 in the search space. This is because in some cases, a seasonal period that is a harmonic of the dominant frequency can lead to better modeling results.
    - The same rationale goes into choosing the `window_length` (number of lags used for modeling) search space as well. 
2. The default number of searches is 10 hyper-parameter combinations.

```python
#### OK, so what search space was used? ----
random_grid = exp.models(internal=True).loc["lr_cds_dt", "Tune Distributions"]
random_grid
```

![image](https://user-images.githubusercontent.com/33585645/139675098-bac21ae8-7cf8-4e0c-8485-4478459ddb03.png)

## Fixed Grid Search

Alternately (instead of performing a random grid search), one may choose to do hyper-parameter tuning using a "fixed" grid search. This can be done in `pycaret` using the `search_algorithm` argument.

```python
#### Tune the model (Fixed Grid Search) ----
tuned_model_grid = exp.tune_model(model, search_algorithm = "grid")
exp.plot_model(tuned_model_grid)
```
![image](https://user-images.githubusercontent.com/33585645/139693855-933224fd-0f86-40bc-8433-0e999a396171.png)

Again, you might be interested in learning about the search space used for this tuning. This can be obtained easily using the `models()` call with `internal=True`. The column `Tune Grid` provides the default fixed grid used for tuning.

```python
#### What search space was used? ----
fixed_grid = exp.models(internal=True).loc["lr_cds_dt", "Tune Grid"]
fixed_grid
```
![image](https://user-images.githubusercontent.com/33585645/139694040-85f4e96a-fded-432a-b783-4eabbdd40302.png)

### Reasons for preferring Random Grid Search over a Fixed Grid Search

- In this case too, six (6) hyper parameters were tuned. The difference is that in case of a fixed grid, all permutations of the hyper-parameters are tried (vs. sampling from a search space in case of a random grid search). Because this can be exhaustive and time consuming, generally the fixed grid searches over a limited space compared to a random grid. This can cause the results of a random grid search to be better than that of a fixed grid search (in general, though not all the time).
- In addition, random grid search is preferred because it ends up not wasting valuable time if a certain hyper-parameter is not helpful in improving the performance. This is explained in this **[wonderful video](https://www.coursera.org/learn/deep-neural-network/lecture/dknSn/tuning-process)** by Andrew Ng.

For these reasons, "random grid search" is the default option in the `pycaret` time series module.

# Suggested Next Reads

- https://github.com/pycaret/pycaret/discussions/1795

**Comment by moezali1** [comment]
PyCaret 4.0 sunset the legacy time-series module (sktime/prophet integration), so this FAQ no longer maps to a supported code path.

If you still need time-series forecasting, use sktime / prophet / statsforecast directly — those projects have surpassed what the PyCaret wrapper offered.

For 4.0 docs, see https://pycaret.org/docs and the cleanup notice at https://github.com/pycaret/pycaret/issues/4177.

Closing as outdated.
